### PR TITLE
refactoring of events, it should be about 4 times faster now

### PR DESCRIPTION
**Pull request by perliedman**
This is an updated version of #2755, originally by @fab1an.

I have merged with current master, resolved conflicts and fixed code style issues.

### CODE REVIEW

File: src/core/Events.js [hunk 1]
```diff
@@ -70,7 +70,7 @@ L.Evented = L.Class.extend({
 
 		if (!types) {
 			// clear all listeners if called without arguments
-			delete this._events;
+			this._events = undefined;
```
Comment (yohanboniface) [inline src/core/Events.js]: Not sure we need that change.

File: src/core/Events.js [hunk 2]
```diff
@@ -90,95 +90,97 @@ L.Evented = L.Class.extend({
 
 	// attach listener (without syntactic sugar now)
 	_on: function (type, fn, context) {
+		this._events = this._events || {};
+
+		/* get/init listeners for type */
+		var typeListeners = this._events[type];
```
Comment (yohanboniface) [inline src/core/Events.js]: Maybe, a more Leaflet way to do that would be something like:

```
var typeListeners = this._events[type] = this._events[type] || {listeners: {}, count: 0};
```

File: src/core/Events.js [hunk 3]
```diff
@@ -188,26 +190,29 @@ L.Evented = L.Class.extend({
...
+			for (var contextId in typeListeners.listeners) {
+				var listeners = typeListeners.listeners[contextId];
 
-				for (i = 0, len = listeners.length; i < len; i++) {
-					listeners[i].fn.call(this, event);
+				// each fn in context
+				for (var i = 0, len = listeners.length; i < len; i++) {
+					var l = listeners[i];
+					if (l.ctx) {
+						l.fn.call(l.ctx, event);
```
Comment (yohanboniface) [inline src/core/Events.js]: `l.fn.call(l.ctx || this, event);`?

File: src/core/Events.js [hunk 4]
```diff
@@ -221,9 +226,8 @@ L.Evented = L.Class.extend({
 	// @method listens(type: String): Boolean
 	// Returns `true` if a particular event type has any listeners attached to it.
 	listens: function (type, propagate) {
-		var events = this._events;
-
-		if (events && (events[type] || events[type + '_len'])) { return true; }
+		var typeListeners = this._events && this._events[type];
+		if (typeListeners && typeListeners.count) { return true; }
```
Comment (yohanboniface) [inline src/core/Events.js]: `if (this._events && this._events[type] && this._events[type].count) { return true; }` ?

File: src/core/Events.js [hunk 5]
```diff
@@ -90,95 +90,97 @@ L.Evented = L.Class.extend({
...
+			typeListeners.listeners[contextId] = listeners;
+		}
 
-				// keep track of the number of keys in the index to quickly check if it's empty
-				events[indexLenKey] = (events[indexLenKey] || 0) + 1;
+		// check if fn already there
+		var found = false;
+		for (var i = 0, len = listeners.length; i < len; i++) {
+			if (listeners[i].fn === fn) {
+				found = true;
```
Comment (yohanboniface) [inline src/core/Events.js]: We may return here and remove `found` check.

File: src/core/Events.js [hunk 6]
```diff
@@ -188,26 +190,29 @@ L.Evented = L.Class.extend({
...
-				listeners = events[type].slice();
+			// each context
+			for (var contextId in typeListeners.listeners) {
+				var listeners = typeListeners.listeners[contextId];
 
-				for (i = 0, len = listeners.length; i < len; i++) {
-					listeners[i].fn.call(this, event);
+				// each fn in context
+				for (var i = 0, len = listeners.length; i < len; i++) {
+					var l = listeners[i];
```
Comment (yohanboniface) [inline src/core/Events.js]: Not sure we need `l` either actually.

File: src/core/Events.js (hunk 1 above)
Comment (fab1an) [inline src/core/Events.js]: Mistake on my part: https://github.com/Leaflet/Leaflet/pull/2755#discussion-diff-15043284

File: src/core/Events.js (hunk 2 above)
Comment (fab1an) [inline src/core/Events.js]: I splitted this up after doing benchmarking, where more lookups are done when using the concise syntax. It's a micro-optimisation though, might not be true for modern browsers anymore.

**Comment by fab1an** [comment]
Thanks to both of you.

**Comment by mourner** [comment]
How does performance look in your benchmarks?

**Comment by mourner** [comment]
Let's follow along in #4709 which includes the changes from this PR.

**Comment by fab1an** [comment]
Should I close my original PR: #2755  as well?

**Comment by perliedman** [comment]
Re-opening since [performance of these changes seem better than #4709](https://github.com/Leaflet/Leaflet/pull/4709#issuecomment-231302559).

**Comment by mourner** [comment]
Let's rebase this on master (cleaning up history with fixups), and then I'm up for merging.

**Comment by perliedman** [comment]
@mourner I rebased on `master`, but did you also mean squash commits?

I've often squashed commits when merging the PR, from GitHub's interface, but I can of course squash on this branch if you prefer it.

**Comment by perliedman** [comment]
@mourner Ok, rebased and squashed it into two commits.

**Comment by fab1an** [comment]
Awesome, I'm happy to see this merged, thanks guys!

**Pull request merged**